java/ql/src/semmle/code/java/dataflow/FlowSources.qll

```diff
@@ -341,3 +341,16 @@ class ExportedAndroidIntentInput extends RemoteFlowSource, AndroidIntentInput {
 
   override string getSourceType() { result = "Exported Android intent source" }
 }
+
+/** Android `PackageInfo` objects that may have come from a hostile application. */
+class AndroidPackageInput extends DataFlow::Node {
+  Type receiverType;
+
+  AndroidPackageInput() {
+    exists(MethodAccess ma |
+      ma.getMethod().getQualifiedName() = "PackageManager.getInstalledPackages"
+    |
+      this.asExpr() = ma
+    )
```

Comment:
Please use `RefType.hasQualifiedName` instead of `Method.getQualifiedName` to match the type, as the latter doesn't include the package name. Also, the field is unused.
```suggestion
  AndroidPackageInput() {
    exists(MethodAccess ma, Method m |
      ma.getMethod() = m and
      m.getDeclaringType().hasQualifiedName("android.content.pm", "PackageManager") and
      m.hasName("getInstalledPackages") and
      this.asExpr() = ma
    )
```